### PR TITLE
Remove the `shots` kwarg from the gradients module

### PR DESCRIPTION
**Pull request by timmysilv**
**Context:**
By the time tapes reach the `gradients` module, they will have shots assigned directly onto them. This means the shots don't need to be passed around while the tapes themselves are being passed around.

**Description of the Change:**
Remove all extraneous references to shot values when they are already on tapes in gradient-related code

**Benefits:**
Less likely to make duplication errors, cleaner code, easier to work with

**Possible Drawbacks:**
Making (and reviewing) this change is ugly.

### CODE REVIEW

**Comment by timmysilv** [comment]
[sc-42756]

File: pennylane/gradients/gradient_transform.py [hunk 1]
```diff
@@ -266,10 +265,10 @@ def _all_zero_grad(tape, shots=Shots(None)):
 
         list_zeros.append(sub_list_zeros)
 
-    if shots.has_partitioned_shots:
+    if tape.shots.has_partitioned_shots:
         if len(tape.measurements) == 1:
-            return [], lambda _: tuple(list_zeros[0] for _ in range(shots.num_copies))
-        return [], lambda _: tuple(tuple(list_zeros) for _ in range(shots.num_copies))
+            return [], lambda _: tuple(list_zeros[0] for _ in range(tape.shots.num_copies))
+        return [], lambda _: tuple(tuple(list_zeros) for _ in range(tape.shots.num_copies))
```
Comment (albi3ro) [inline pennylane/gradients/gradient_transform.py:271]: Completely optional, but we could leave this code cleaner than we found it and turn the lambas into normal functions.
```suggestion
           
           num_copies = tape.shots.num_copies
           def zero_grad(*_):
               return tuple(list_zeros[0] for _ in range(num_copies)

            return [], zero_jacobian
            
        num_copies = tape.shots.num_copies
        def zero_grad(*_):
           return tuple(list_zeros for _ in range(num_copies)
        return [], zero_grad
```

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -160,6 +160,9 @@ array([False, False])
 * `transmon_drive` is updated in accordance with [1904.06560](https://arxiv.org/abs/1904.06560). In particular, the functional form has been changed from $\Omega(t)(\cos(\omega_d t + \phi) X - \sin(\omega_d t + \phi) Y)$ to $\Omega(t) \sin(\omega_d t + \phi) Y$.
   [(#4418)](https://github.com/PennyLaneAI/pennylane/pull/4418/)
 
+* The gradients module no longer needs shot information passed to it explicitly, as the shots are on the tapes.
+  [(#4448)](https://github.com/PennyLaneAI/pennylane/pull/4448)
+
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:165]: should be under breaking change.

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -205,6 +205,9 @@ array([False, False])
 * The Pauli-X-term in `transmon_drive` has been removed in accordance with [1904.06560](https://arxiv.org/abs/1904.06560)
   [(#4418)](https://github.com/PennyLaneAI/pennylane/pull/4418/)
 
+* The gradients module no longer needs shot information passed to it explicitly, as the shots are on the tapes.
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:208]: Please could we also list this as a breaking change (removing shots from gradient transforms)

**Review by eddddddy — APPROVED** [review]
Looks good 🎉

File: pennylane/interfaces/tensorflow.py [hunk 4]
```diff
@@ -94,14 +94,14 @@ def _to_tensors(x):
...
-def _res_restructured(res, tapes, legacy_shots: Shots):
+def _res_restructured(res, tapes):
     """
     Reconstruct the nested tuple structure of the output of a list of tapes
     """
     start = 0
     res_nested = []
     for tape in tapes:
-        tape_shots = legacy_shots or tape.shots or Shots(1)
+        tape_shots = tape.shots or Shots(1)
```
Comment (dwierichs) [inline pennylane/interfaces/tensorflow.py:104]: This seems fine because it is unrelated, but I was actually surprised to see this single-shot logic here for cases like `tape.shots=None`.

File: pennylane/interfaces/jax.py [hunk 5]
```diff
@@ -440,12 +440,7 @@ def _execute_bwd(
     # pylint: disable=unused-variable
     # Copy a given tape with operations and set parameters
 
-    if isinstance(device, qml.devices.experimental.Device):
-        # cant test until we integrate device with shot vector
-        has_partitioned_shots = tapes[0].shots.has_partitioned_shots
-        jvp_shots = None
-    else:
-        has_partitioned_shots = jvp_shots = device.shot_vector
+    has_partitioned_shots = tapes[0].shots.has_partitioned_shots
```
Comment (dwierichs) [inline pennylane/interfaces/jax.py:443]:
```suggestion
    # assumes all tapes have the same shot vector
    has_partitioned_shots = tapes[0].shots.has_partitioned_shots
```

File: pennylane/interfaces/tensorflow.py [hunk 6]
```diff
@@ -332,13 +332,7 @@ def execute(tapes, device, execute_fn, gradient_fn, gradient_kwargs, _n=1, max_d
...
     params_unwrapped = []
 
-    if isinstance(device, qml.devices.experimental.Device):  # pragma: no-cover
-        # assumes all tapes have the same shot vector
-        has_partitioned_shots = tapes[0].shots.has_partitioned_shots
-        vjp_shots = legacy_shots = None
-    else:
-        has_partitioned_shots = vjp_shots = device.shot_vector
-        legacy_shots = Shots(device.shot_vector or 1)
+    has_partitioned_shots = tapes[0].shots.has_partitioned_shots
```
Comment (dwierichs) [inline pennylane/interfaces/tensorflow.py:335]:
```suggestion
    # assumes all tapes have the same shot vector
    has_partitioned_shots = tapes[0].shots.has_partitioned_shots
```

File: doc/releases/changelog-dev.md (hunk 3 above)
Comment (timmysilv) [reply doc/releases/changelog-dev.md:208]: it's been moved to the breaking changes section

**Review by dwierichs — COMMENTED** [review]
Thanks so much for this change @timmysilv ! :) 
Very nice to see this kwarg go :tada: 
I just had a few very minor comments. I'll to a second round of trying to find leftover code bits and such, but the changes look good to me. :)

File: pennylane/gradients/finite_difference.py [hunk 7]
```diff
@@ -473,7 +463,7 @@ def _single_shot_batch_result(results):
         return tuple(tuple(elem) for elem in grads_reorder)
 
     processing_fn = functools.partial(
-        _processing_fn, shots=shots, single_shot_batch_fn=_single_shot_batch_result
+        _processing_fn, shots=tape.shots, single_shot_batch_fn=_single_shot_batch_result
```
Comment (dwierichs) [inline pennylane/gradients/finite_difference.py:468]: [Commenting here because the definition line of `_processing_fn` wasn't touched in the PR and can't be commented]

I think we could move to have `_processing_fn` only take positional arguments, because it is a very particular function only used once here and once in the spsa gradient, at the moment.
Then we can get rid of the dummy `shots=Shots(None)` which is executed whenever loading PL, as far as I am aware.

**Review by dwierichs — APPROVED** [review]
Couldn't find anything else, except for one tiny nitpicky thing :)
Definitely not blocking.

File: pennylane/gradients/finite_difference.py (hunk 7 above)
Comment (timmysilv) [reply pennylane/gradients/finite_difference.py:468]: done!
